### ISSUE TITLE
[Cohort-1 Task 23] 降低 colocate 权重同步切换开销

### ISSUE THREAD

**Issue by xiaoh1024** [redai-infra/Relax#240]
## 背景

Qwen3-4B colocate recipe 在每轮 Actor 训练后，当前会先 offload Actor，再将 CPU 备份权重加载回 GPU，完成 Megatron-to-HF 转换和 SGLang 权重更新。基线中 `update_weights` 平均耗时约 5 秒，占平均 step time 的 2.8%，是每轮 GPU 角色切换中的一段固定开销。

计划增加默认关闭的 `--colocate-live-weight-sync`：显存允许时，直接从训练后的 GPU Actor 参数同步 SGLang 权重，再 offload Actor 并恢复 rollout KV cache；显存不足或最终 checkpoint step 自动使用现有 CPU-backed 路径。

```text
Actor 训练完成
  → 全 rank 显存检查
  → 从 live Actor tensor 更新 SGLang
  → offload Actor
  → 恢复 rollout KV cache
```

## 代码基线

- 上游仓库：https://github.com/redai-infra/Relax
- 开发仓库：https://github.com/xiaoh1024/Relax
- 基线 commit：`0bc99af8dd39de8fd99c588a98b3f3a463bc818c`
- 开发分支：`perf/task23-colocate-weight-sync`

## 环境与配置

- 镜像：`ghcr.io/redai-infra/relaxrl:dev-20260715-8325919e@sha256:3fa8ce578acda6c829b83016bde42c38fa892681e4f36ca330f545616fe578e2`
- Python 3.12.3；PyTorch 2.11.0+cu129；CUDA 12.9；SGLang 0.5.12.post1；Ray 2.56.0；Transformers 5.6.0
- 单节点：8 × NVIDIA L20Z 80 GiB，128 vCPU，1024 GiB 内存
- 模型：`Qwen/Qwen3-4B@1cfa9a7208912126459214e8b04321603b3df60c`
- 数据：`zhuzilin/dapo-math-17k@2e65612930298bde4c5d58fd97b3f23a483aaff9`
- 评测：`zhuzilin/aime-2024@1c625e328db94ec7ef7ff169016b097c468d60b9`
- Recipe：`scripts/training/text/run-qwen3-4B-8xgpu.sh`
- GRPO，training TP2/DP4，rollout TP8，rollout batch size 32，`n=8`，global batch size 128，max response length 8192，temperature 1，seed 1234

## 基线结果

两次运行均为 21 个 rollout；0–1 为 warmup，2–19 纳入统计，20 包含 eval、不纳入主指标。

| Run | Total tokens/s | Response tokens/s | Step time | Weight update | GPU util | Peak VRAM |
|---|---:|---:|---:|---:|---:|---:|
| base-1 | 9570.19 | 9352.09 | 177.49 s | 5.07 s | 87.41% | 71.34 GiB |
| base-2 | 9594.72 | 9378.99 | 178.26 s | 4.96 s | 87.52% | 70.07 GiB |
| mean | 9582.46 | 9365.54 | 177.88 s | 5.02 s | 87.47% | max 71.34 GiB |

## 改动范围

- `relax/backends/megatron/actor.py`：调整同步、offload 和 KV 恢复顺序；负责一致回退。
- `relax/backends/megatron/weight_update/live_weight_sync.py`：显存预算和全 rank 决策。
- `relax/backends/megatron/weight_update/update_weight_from_tensor.py`：支持直接读取 live Actor tensor。
- `relax/utils/arguments.py`、`scripts/training/text/run-qwen3-4B-8xgpu.sh`：增加默认关闭的实验开关。
- `tests/backends/megatron/weight_update/test_live_weight_sync.py`：覆盖显存计算、全 rank 决策、执行顺序和异常清理。

范围限定为同步 colocate、Megatron bridge、Actor + rollout、PP1/EP1、非 LoRA；不涉及训练算法或 SGLang kernel。

## 运行命令

Baseline：

```bash
export MODEL_DIR=/path/to/models
export DATA_DIR=/path/to/data
export EXP_DIR=/path/to/output/base
export NUM_ROLLOUT=21
bash scripts/training/text/run-qwen3-4B-8xgpu.sh
```

Candidate：

```bash
export MODEL_DIR=/path/to/models
export DATA_DIR=/path/to/data
export EXP_DIR=/path/to/output/live-weight-sync
export NUM_ROLLOUT=21
export COLOCATE_LIVE_WEIGHT_SYNC=1
bash scripts/training/text/run-qwen3-4B-8xgpu.sh
```

## 验收口径

- `perf/step_token_per_s` 相对两次 baseline 均值提升至少 1%。
- `perf/update_weights_time` 降低至少 70%，`sleep + update_weights + wake_up` 降低至少 30%。
- 有效样本和 token 口径一致，loss/reward 无异常；无 OOM 或未解释异常。
- 默认配置行为不变；live tensor 与 CPU-backed 路径的转换结果和 weight version 一致。

请维护者确认该优化方向、支持范围和验收口径是否合适。

**Comment by xiaoliang0601** [comment]
❌ 不通过。

收益略小，不是训练过程中的 hot path。
